### ISSUE TITLE
Missing .NET Standard from Existing Project Target Framework Options

### ISSUE THREAD

**Issue by pracplayopen** [mono/monodevelop#9515]
Was trying to retarget existing project to .net standard in monodevelop.

Could not do it.  Then saw this very similiar issue:  https://github.com/mono/monodevelop/issues/6146

However did not seem to apply.  

Was already running newer versions (monodevelop 7.3 mono 5.18 and dotnet 3.1/2.2), still did not see options.    Even tried running latest monodevelop for ubuntu xenial stable (7.8.4 build 2), and still could not change an existing project to target .net standard.

Am I misunderstanding something about .net standard?   As an experiment I tried creating a new project, and in both monodevelop versions (7.3 and 7.8) .net standard is a project template option.   (Was a little confused which was being referred to in issue #6164)

1. Is not being able to target .NET standard on existing projects expected behavior?
2. if so, is this a msft thing or just trying to force a good practice on people?
3. assuming this is expected, is there a way i can create a .net standard library that shares all files from the framework library?  (i already know everything in this library will run in .netcore with no issues)

Thanks for any assistance or clarification.

**Comment by mrward** [comment]
Retargeting the project from what which framework?

Currently the only thing supported is to convert a PCL project to .NET Standard.

Have you looked at James Montemagno's - [How to convert a PCL library to .NET Standard](https://montemagno.com/how-to-convert-a-pcl-library-to-net-standard-and-keep-git-history/)?

**Comment by pracplayopen** [comment]
ah ok, no these are involving .net-framework library (bcl), not portable.

is there a way to create a new .net standard project that includes all the files from framework proj?

suppose I could symlink them?  never tried that en-mass, not sure if will cause weird problems.

i have seen the 'shared project' in monodevelop, but few docs for it so hard to understand.

Can shared projects be targeted at .netstanard and .net framework at same time?

do 'shared project' builds well across all of xbuild/msbuild/dotnet-build.

thanks for your response

**Comment by mrward** [comment]
I would try creating a .NET Standard project and then, as in James Montemagno's video, copy that over the top of the existing .csproj and go from there. Backup the original project first though 😄

SDK style projects have various default includes for files so you do not need to specify each file in the project.

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
</Project>

```

You will likely need to add PackageReferences for any NuGet packages you are using. Either manually by editing the csproj file or just installing them again in the IDE.

Shared projects do not target any framework they are just containers for files. The project that references the shared project determines the target framework.

Note that SDK style projects support multi-targeting. So you can have it target both .NET Standard and .NET Framework if you need to.

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net462</TargetFrameworks>
  </PropertyGroup>
</Project>

```

**Comment by pracplayopen** [comment]
thanks for tips.   

the video you're referencing i think is this one: https://www.youtube.com/watch?v=F0VWCmj2adw

Pretty sure i'll need to build both old framework versions and new ones together for some time.  

Don't think I can use multi-targetting bc believe that only works w/.net 4.6 or so or higher, but we still have 4.0 needs in some places.

in case anybody else has this issue, what I did was create the new .net standard projects as you suggested, then synced up project dependencies.  

you were right the project files don't need specific file itemizations here (almost a reason to upgrade on it's own! :)

anyways that let me write a little script, which given a src and dst copies all the cs files to the standard project which were/are being built by the framework project.    This worked:

```
#!/bin/sh

src=$1
dst=$2
help="sync-dotnet.sh <src> <dst>"
if [ ! -d "$src" ]; then
        echo "$help">&2
        echo "error, no source project dir at: $src">&2
        exit 1
fi

if [ ! -d "$dst" ]; then
        echo "$help">&2
        echo "error, no dest project dir at: $src">&2
        exit 2
fi
proj=$(find "$src" -type f -iname '*.csproj'|head -1)
if [ -z "$proj" ] || [ ! -f "$proj" ]; then
        echo "error, unable to find csproj file $proj in $src">&2
        exit 3
fi
inclfs=$(egrep -i "Include=\"[a-z0-9_\-\.]+.cs\"" "$proj"|sed 's/Compile Include=//g'|sed 's/[<>\/" ]//g')
skipfs="AssemblyInfo.cs _svntemplate.cs"

csps=$(find "$src" -type f -iname '*.cs')

errc=0
for csp in $csps; do
        fn=$(basename "$csp")
        # check that should be included
        if echo "$skipfs"|grep -q "$fn"; then
                echo "hard-skip requested for: $fn...">&2
                continue;
        fi
        if ! echo "$inclfs"|grep -q "$fn"; then
                echo "skipping non-built file: $fn...">&2
                continue
        fi
        dstp="$dst/$fn"
        cp -fv "$csp" "$dstp"
        if [ $? != 0 ]; then
                echo "error copying $csp to $dstp">&2
                errc=$((errc+1))
        fi
done
exit $errc

```

Now should be able to add as pre/post build event, so it keeps things manageable.

While doing this I also ran into common incompatability issue between mono 5.18 msbuild+nuget and newer version of dot-net sdks, which I could only fix by uninstalling dotnet-sdk-3.1

https://rider-support.jetbrains.com/hc/en-us/articles/360004180039
https://github.com/dotnet/core/issues/2460

thanks for your help

**Comment by GeorgeS2019** [comment]
@pracplayopen  I could not get MONO msbuild to compile sdk-style project. Can someone share links where I can follow to make it works?
